clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -346,7 +346,11 @@ public void onFailure(RuntimeException e) {
                             FetchSessionHandler handler = sessionHandler(fetchTarget.id());
                             if (handler != null) {
                                 handler.handleError(e);
-                                handler.sessionTopicPartitions().forEach(subscriptions::clearPreferredReadReplica);
+                                // Make sure to filter out topic partitions that are not part of the assignment
+                                // anymore when the request fails.
+                                handler.sessionTopicPartitions().stream()
+                                    .filter(subscriptions::isAssigned)
+                                    .forEach(subscriptions::clearPreferredReadReplica);
```

Comment:
What would happen if the partition is removed from the subscriptions between these two lines? It seems to me that we could face the same issue.

It may be better to push this check into `clearPreferredReadReplica`. Something like this:

```
    public synchronized Optional<Integer> clearPreferredReadReplica(TopicPartition tp) {
        final TopicPartitionState topicPartitionState = assignedStateOrNull(tp);
        if (topicPartitionState == null) {
            return Optional.empty();
        } else {
            return topicPartitionState.clearPreferredReadReplica();
        }
    }
```

Reply:
Good point! I was thinking the `Fetcher` object monitor protects us, but you are right, 
 it doesn't.